src/txmempool.h

```diff
@@ -847,7 +847,14 @@ class CTxMemPool
         friend class CTxMemPool;
     };
 
-    std::unique_ptr<CTxMemPoolChangeSet> GetChangeSet() EXCLUSIVE_LOCKS_REQUIRED(cs) { return std::make_unique<CTxMemPoolChangeSet>(this); }
+    std::unique_ptr<CTxMemPoolChangeSet> GetChangeSet() EXCLUSIVE_LOCKS_REQUIRED(cs) {
+        Assume(m_changeset == nullptr);
+        auto ret = std::make_unique<CTxMemPoolChangeSet>(this);
+        m_changeset = ret.get();
+        return ret;
+    }
+
+    CTxMemPoolChangeSet* m_changeset GUARDED_BY(cs){nullptr};
```

Comment:
This only exists for sanity checking that we don't have more than one changeset at a time.  We could just as easily replace this with a `bool`; I made it a pointer mostly for debugging purposes, if the `Assume()` were ever hit.

The idea behind the `unique_ptr` is that I wanted all the state in the ChangeSet to magically go away whenever the object goes out of scope, so that I didn't have to explicitly clean up state everywhere that a transaction might fail validation (which seems like a lot of places).  In this PR, there's not a ton of state, but I expect that when we implement cluster mempool that we'll be creating similar staging data structures within the `TxGraph` layer that will exist one level below the mempool (for computing clusters and feerate diagrams for candidate transactions), and I thought this mechanism of ending a staging when the changeset goes out of scope would be an easy way to manage things at that layer as well.  So in the future you could imagine that in the `CTxMemPoolChangeSet` constructor, we might initiate a staging in the underlying `TxGraph`; in the mempool `Apply()` function we might also apply the staging to the `TxGraph`; and in the `CTxMemPoolChangeSet` destructor we might abort a staging in the underlying `TxGraph`.

The paradigm I'm looking for is that you can have at most one changeset at a time, and when there's a changeset outstanding that serves as a lock on the mempool so that you can't otherwise add/remove things (except via the changeset) until the changes are either applied or aborted, and the changeset is the destroyed.  However if there's a better way to implement that than what I tried here, I'm certainly open to suggestions.

Reply:
Done.